kernel/src/history_manager/search.rs

```diff
@@ -0,0 +1,253 @@
+//! This module defines the [`binary_search_by_key_with_bounds`] utility method. This can be used to
+//! search over sorted slices of values to find the greatest lower or least upper bounds.
+use std::cmp::Ordering;
+use std::error::Error;
+use std::fmt::Debug;
+
+use delta_kernel_derive::internal_api;
+
+/// Defines the type of bound to return from a binary search operation.
+///
+/// For a search operation over `values` using key `key`:
+/// * [`Bound::LeastUpper`] - Finds the smallest index `i` such that `values[i] >= key`.
+///   This represents the first element greater than or equal to the search key.
+///
+/// * [`Bound::GreatestLower`] - Finds the largest index `i` such that `values[i] <= key`.
+///   This represents the last element less than or equal to the search key.
+#[internal_api]
+#[allow(unused)]
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum Bound {
+    LeastUpper,
+    GreatestLower,
+}
+
+#[internal_api]
+#[allow(unused)]
+#[derive(Debug)]
+pub(crate) enum SearchError<T: Error> {
+    OutOfRange,
+    KeyFunctionError(T),
+}
+
+/// Performs a binary search on a slice to find the greatest lower or least upper bound.
+/// This function performs a binary search with the following properties:
+///
+/// 1. For [`Bound::LeastUpper`], find the smallest index i such that values[i] >= key
+/// 2. For [`Bound::GreatestLower`], find the largest index i such that values[i] <= key
```

Comment:
doctest complains about "unresolved link to `i`" -- not sure how to fix that? I guess with backticks?
```suggestion
/// 1. For [`Bound::LeastUpper`], find the smallest index `i` such that `values[i] >= key`
/// 2. For [`Bound::GreatestLower`], find the largest index `i` such that `values[i] <= key`
```